Ghidra/Processors/8048/data/languages/8048.slaspec

```diff
@@ -0,0 +1,349 @@
+# sleigh specification file for Intel 8048
+
+
+# Do not take BS into account when decompiling
+@define SINGLE_REGISTER_BANK
+
+# Treat R0-R7 as not memory mapped (implies SINGLE_REGISTER_BANK)
+@define INTERNAL_REGISTERS
+
+
+@ifdef INTERNAL_REGISTERS
```

Comment:
Well, the reason I want to offer these different modes is that the disassembly becomes gradually better when you change between them.  Consider the interrupt service routine at `CODE:0141` in the Compis keyboard ROM.  This is a very simple function to output a short square wave on a GPIO pin (it's the "bell" operation).  It also sets BS explicitly.
With `INTERNAL_REGISTERS` and `SINGLE_REGISTER_BANK` both disabled, this becomes
```
void FUN_CODE_0141(void)

{
  char *pcVar1;
  char cVar2;
  
  cVar2 = '\x01';
  BANK1_R0 = 0x80;
  do {
    BUS = BUS | 0x20;
    FUN_CODE_014f();
    BUS = BUS & 0xdf;
    FUN_CODE_014f();
    pcVar1 = (char *)(cVar2 * '\x18');
    *pcVar1 = *pcVar1 + -1;
  } while (*pcVar1 != '\0');
  return;
}
```
which is accurate but looks rather horrible.  With `SINGLE_REGISTER_BANK` defined, it becomes
```
void FUN_CODE_0141(void)

{
  setBank(1);
  BANK0_R0 = -0x80;
  do {
    BUS = BUS | 0x20;
    FUN_CODE_014f();
    BUS = BUS & 0xdf;
    FUN_CODE_014f();
    BANK0_R0 = BANK0_R0 + -1;
  } while (BANK0_R0 != '\0');
  return;
}
```
(which is techniclly wrong because `BANK1_R0` is used, not `BANK0_R0`) and with `INTERNAL_REGISTERS` it becomes
``` 
void FUN_CODE_0141(void)

{
  char cVar1;
  
  setBank(1);
  cVar1 = -0x80;
  do {
    BUS = BUS | 0x20;
    FUN_CODE_014f();
    BUS = BUS & 0xdf;
    FUN_CODE_014f();
    cVar1 = cVar1 + -1;
  } while (cVar1 != '\0');
  return;
}
```
which looks much nicer.  So the modes allow a trade-off between accuracy and actually readable code.

I don't mind changing the default mode to something else, but removing the INTERNAL_REGISTERS mode completely seems a waste.  What is this SleighEditor that is causing you so much trouble and why can't it handle slaspecs which conform to the SLEIGH specification?

Reply:
Eh, I meant "decompilation", not "disassembly", obviously. :smile:

Reply:
In my quick look, I could see the complication with the mapping of registers with BS.  I didn't want to complicate the review knowing I was going on vacation, but had wanted to give you feedback before I left.  Apologize for the half answer.

Memory mapped registers can be difficult to deal with, and can have effects in the decompiler and for constant propagation.  The changes I had in mind to deal with it have a few complications.

The problem is you don't know the BS.  You can assume it's value, and the decompiler/disassembly will figure out the value if some routine changes the BS.  I guess the real question is, "is it worth it?".  Does code exist that uses the BS register banking?  Will anyone ever emulate it?  If so, then I think it can be made to work by:

The first change that might solve your issues is adding an assumption of BS set to 0 in the pspec.  Then change the attach to be by name, or just define labels in the pspec for the memory mapped register locations as you have done.  You might not need to do much more to get good decompilation.

'''
<programcounter register="PC">
<context_data>
    <tracked_set space="CODE">
        <set name="BS" val="0"/>
    </tracked_set>
</context_data>
'''
This will allow setting of the BS to 0 or 1 over an area of code if the value is static.  Even if the BS changes within a routine, this will work fine.  If the value changes and shouldn't be assumed to '0' because it matters to the RE, then there is a more complicated method.

If the register is set to various values, and might take on 0 or 1 in the same routine based on different calls to it then something different is necessary.  If the BS is not assumed, a segment()  trick on the R registers, similar to what the 16-bit x86 does with the segment registers, can make the decompiler work correctly.  Something like below is added

To the 8048.cspec
'''
  <segmentop space="INT_MEM" userop="segment" baseinsize="1" innerinsize="1" farpointer="yes">
    <baseop code="INT_ZEXT"/>
    <baseop code="INT_MULT" value="24"/>
    <innerop code="INT_ZEXT"/>
    <constresolve>
      <register name="BS"/>
    </constresolve>
  </segmentop>
'''

To the 8048.slaspec:

'''
define pcodeop segment; # Define special pcodeop that calculates banked registers

...
macro regbank(r) {
        r = segment(BS,r);
}

'''

I'm sure my suggested edits aren't perfect and will need some modification. The first method is probably worth adding to keep the code consistent with few ifdefs.  It allows setting the BS register as long as it is set across an area consistently, or changed within a function correctly.

Emulation will work in both cases.  If the BS is not assumed, then constant propagation will most likely not work quite right, but actually could be made to work if it becomes an issue.

Reply:
The sleigh editor will be included in 9.1.  We've merged the code into master, but it currently must be built or run in eclipse to use.  We'll be adding a fully build eclipse plugin soon, and the release will contain a built eclipse plugin like the GhidraDev eclipse plugin.

The features are mentioned here:
https://github.com/NationalSecurityAgency/ghidra/wiki/files/recon2019.pdf

The code and the build docs are here:
https://github.com/NationalSecurityAgency/ghidra/tree/master/GhidraBuild/EclipsePlugins/SleighEditor

Reply:
Thanks for the suggested improvements, I'll try them out when I get around to it.  Are the `tracked_set` and `segmentop` entities (or indeed any of the `cspec` file) documented anywhere?  I couldn't find anything relevant in `GhidraDocs`, so I basically cargo-culted the whole cspec from specs for similar processors...

As for the expected usage of BS, I think 99% of all code would do the same thing as the Compis keyboard code; set BS=1 in interrupt handlers to avoid the need of register save, and leave it at the default BS=0 in main code.  Note that the same subroutine might be called both from interrupt and main code, and thus with different BS, such as the delay function `FUN_CODE_014f` in the example.  In such a case, any memory mapped registers clobbered (`R1` in the case of `FUN_CODE_014f`) will almost certainly have local lifetime, so as long as input and output parameters are not affected guessing RS wrong is unlikely to cause any issues.  For this reason, I set the default implementation to one which which does _not_ take RS into account; _usually_ it won't have any affect on RE. :smile:

Reply:
Regarding the editor, I don't think the use of `#` as a literal instead of a comment character in the display section violates context-freeness.  That part of the language is certainly within LR(1), even though a naïve implementation based on that fact might not be the most efficient.  As for the use of `@define` with an implied value, that should be _very_ straightforward to handle.

If the code is available now I might take a stab at fixing this, but I make no promises because now _my_ vacation is coming to an end.  :smile_cat:

Reply:
There .cspec and .pspec are evolving, and there are some docs in the works, but they aren't complete.  I usually look for other processors that have the same idiosyncrasies.  This particular usage to handle segmenting as only done in the 16-bit x86.  The other processors didn't seem to need it most of the time as the main registers usually aren't done this way, or they can't be known until run-time.  So if a particular function just uses the registers without knowledge of what bank they are in, the decompiler should work fine.

You might have some issues passing the r0-r7 registers as parameters, since it won't know which register is correct.  There may be other thorns.

Reply:
The grammer for the editor, and the grammer for the full sleigh compiler are two different implementations.  The XTEXT parser might be able to be fixed to handle "#" in the display piece.  I haven't found it worth it, as you can just put "" around it.  The @define with an empty value is an issue since normally a '\n' is considered whitespace, but I suppose could be handled with a sub-lexer just like the "#".  I'll take a look.

The use of macros anywhere you want causes major issues to the parser.  It works if you always know the value and can look it up.  However feeding the value of a macro into XTEXT mid-parse is non-trivial.  IMHO cavalier use of macro expansion leads to bad coding practices anyway.  So the choice was to allow it in restricted locations.  The one in the 8048 should be allowed, but with the refactor as noted above, it may not be necessary.

That said allowing a macro in the name of the address space is easily supported.